### ISSUE TITLE
unlisted deps are not pruned automatically if two-package.json structure not used

### ISSUE THREAD

**Issue by yashagarwal1411** [electron-userland/electron-builder#770]
* **Version**: 7.3.0

* **Target**: macOS

Code signing failing only when asar option is set to true. When set as false code signing works fine and dmg and zip files are created.

Node version - v6.2.2
npm version - 3.9.5
My System - OS X El Capitan 10.11.4

Logs:

```
electron-osx-sign Signing... /opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libffmpeg.dylib +2s
  electron-osx-sign Signing... /opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libnode.dylib +434ms
  electron-osx-sign Signing... /opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Resources/crashpad_handler +667ms
  electron-osx-sign Signing... /opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework +401ms
Error: Command failed: codesign --sign Developer ID Application: Eltropy Inc (L7YA27678X) --force /opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework
/opt/eltropy/cello/dist/mac/Eltropy Sync.app/Contents/Resources/app.asar.unpacked/app/node_modules/electron-prebuilt/dist/Electron.app/Contents/Frameworks/Electron Framework.framework: bundle format unrecognized, invalid, or unsuitable

    at ChildProcess.exithandler (child_process.js:207:12)
From previous event:
    at /opt/eltropy/cello/node_modules/electron-osx-sign-tf/index.js:443:18
    at ChildProcess.exithandler (child_process.js:198:7)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
From previous event:
    at /opt/eltropy/cello/node_modules/electron-osx-sign-tf/index.js:437:27
From previous event:
    at signApplicationAsync (/opt/eltropy/cello/node_modules/electron-osx-sign-tf/index.js:395:6)
    at /opt/eltropy/cello/node_modules/electron-osx-sign-tf/index.js:564:14
From previous event:
    at Function.signAsync (/opt/eltropy/cello/node_modules/electron-osx-sign-tf/index.js:556:6)
    at MacPackager.<anonymous> (/opt/eltropy/cello/node_modules/electron-builder/src/macPackager.ts:183:21)
    at next (native)
From previous event:
    at tsAwaiter (/opt/eltropy/cello/node_modules/electron-builder/src/util/awaiter.ts:10:47)
    at MacPackager.sign (/opt/eltropy/cello/node_modules/electron-builder/src/macPackager.ts:98:75)
    at nonMasPromise.doPack.then (/opt/eltropy/cello/node_modules/electron-builder/src/macPackager.ts:78:26)
From previous event:
    at MacPackager.<anonymous> (/opt/eltropy/cello/node_modules/electron-builder/src/macPackager.ts:78:10)
    at next (native)
From previous event:
    at tsAwaiter (/opt/eltropy/cello/node_modules/electron-builder/src/util/awaiter.ts:10:47)
    at Object.build (/opt/eltropy/cello/node_modules/electron-builder/src/builder.ts:181:52)
    at Object.<anonymous> (/opt/eltropy/cello/node_modules/electron-builder/src/build-cli.ts:27:28)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (node.js:348:7)
    at startup (node.js:140:9)
    at node.js:463:3
```

My build package.json:

```
{
  "devDependencies": {
    "electron-builder": "7.3.0",
    "electron": "1.4.0"
  },
  "build": {
    "asar": true,
    "appId": "com.eltropy.syncapp",
    "productName": "Eltropy Sync",
    "category": "public.app-category.business",
    "win": {
      "iconUrl": "https://s3-us-west-2.amazonaws.com/sync-app/assets/ico.ico",
      "loadingGif": "build/installer.gif",
      "icon": "build/ico_256.ico",
      "setupIcon": "build/ico.ico"
    },
    "dmg": {
      "title": "EltropySync",
      "icon": "build/icon.icns",
      "background": "build/mac_install.png",
      "contents": [
        {
          "x": 450,
          "y": 320,
          "type": "link",
          "path": "/Applications"
        },
        {
          "x": 180,
          "y": 320,
          "type": "file"
        }
      ]
    }
  },
  "scripts": {
    "postinstall": "install-app-deps",
    "start": "electron ./app",
    "dist": "build -w --ia32",
    "distm": "build -m"
  }
}
```

To test further I cloned a sample electron app repo - 
https://github.com/develar/onshape-desktop-shell

This initially had "asar": false which I changed to true. And then I tried building the app. Everything worked fine including code signing and build.

**Comment by develar** [comment]
`electron-prebuilt` must be in your `devDependencies` — please check app package.json

**Comment by develar** [comment]
Could you please attach your project? I am very interested why `electron-prebuilt` is not excluded and sample project will help me a lot.

**Comment by yashagarwal1411** [comment]
Hi @develar I am using the two package.json structure. I tried adding electron-prebuilt to my app's package.json file but it didn't help, I am still getting the same issue.

I cannot attach the project as it is a private project. But I can probably try trimming down the project so that it doesn't contain any business logic and then if the issue is still coming, I will attach it here.

In the meanwhile, this is my project structure-

<img width="178" alt="screen shot 2016-09-23 at 12 32 18 pm" src="https://cloud.githubusercontent.com/assets/5121955/18776884/cdf0d2ee-8189-11e6-9521-4c10a3011ee5.png">

My app's package.json

`{
  "name": "cello",
  "productName": "Eltropy Sync",
  "version": "1.9.19",
  "description": "App for syncing files between local system and Eltropy web-app",
  "main": "./main.js",
  "scripts": {
    "start": "electron ."
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/eltropy/cello.git"
  },
  "keywords": [
    "Eltropy",
    "Sync"
  ],
  "author": "Eltropy Inc.",
  "dependencies": {
    "auto-launch": "^5.0.1",
    "bower": "^1.7.9",
    "electron-debug": "^1.0.1",
    "fs-extra": "^0.30.0",
    "fswin": "^2.16.501",
    "log4js": "^0.6.38",
    "mkdirp": "^0.5.1",
    "request": "^2.74.0",
    "winattr": "^1.1.0"
  },
  "devDependencies": {
    "electron-prebuilt": "1.4.1"
  }
}
`

**Comment by develar** [comment]
@yashagarwal1411 Sorry — I mean that if and only if there is a `electron-prebuilt` dependency in your package.json — you should check that it in the `devDependencies`

In any case — please try version `7.8.0`

**Comment by yashagarwal1411** [comment]
I updated electron-builder to 7.8.0 but still the same issue.

**Comment by develar** [comment]
Please set env `DEBUG=electron-builder` and attach build log.

**Comment by yashagarwal1411** [comment]
[log.txt](https://github.com/electron-userland/electron-builder/files/489624/log.txt)

I have attached the build log. But I think now that issue is with my system as I cloned the repo on some of my colleagues macs and started the build and it worked just fine with code signing and produced the necessary dmg and zip files.

Thank you so much for your quick replies on this.

**Comment by develar** [comment]
@yashagarwal1411 Ok... I have no idea. Please try to remove all `node_modules` (in the root of the project and app) and install `npm install` again.

**Comment by yashagarwal1411** [comment]
I did that but still the same issue. Anyways its not a priority for me anymore, I can always use my colleagues computers whenever I want to build. I think we can close this issue and reopen it if someone else faces it.

**Comment by develar** [comment]
Please find directory `electron-prebuilt` in your project. Where is it located?

**Comment by yashagarwal1411** [comment]
Hi @develar I got the issue and its working fine for me now. I feel naive for not figuring this out myself.

I searched for electron-prebuilt and it was in /cello/app/app/node_modules/ folder. I have no idea why the lower app folder contained node_modules and how it got created (It's just a directory for storing angular app files and doesn't have a package.json). But when I deleted this node_modules folder and tried building everything worked fine.

Thanks a lot in helping me debug and fix this issue.
